### DISCUSSION TITLE
Zeek Package Installation

### DISCUSSION THREAD

**Discussion by jlewis200** [Security-Onion-Solutions/securityonion#7728] [Unsupported Versions]
What is the preferred way of installing Zeek packages on SO 2.3.110?  There is documentation for the installation of Zeek scripts, but I did not see anything about installing Zeek packages.  The packages I'm interested in contain scripts and a "plugin" component as Zeek calls them.  I can extract the scripts and deploy those as described in the documentation, but the plugin component would be missing.  On a native install of Zeek I would use the "zkg" package manager, but it appears that Zeek runs inside a container on SO.

**Comment by dougburks** [comment]
Can you share which Zeek packages you are interested in so that we can take a look at them?

**Reply by jlewis200** [reply]
Hi Doug. Here's an example https://github.com/cisagov/icsnpp-enip. There's a script component and also a plugin component to this package. I've tried a few different experiments to see if I could get this running on SO. One of them was jumping into the Zeek container and building/installing from there. The problem I ran into is the package requires cmake>=3.0 and gcc>=9.0, while 2.8/4.8 are the latest available from the CentOS7 repos. I've been tossing around the idea of providing my own Zeek Docker image, but I have doubts as to stability with SO updates and/or my custom image being overwritten.

**Reply by archery713** [reply]
I know this is a few months old but I am curious if you found a solution. I am trying to install the exact same script
Would installing SO on an Ubuntu based platform work around this issue?

**Reply by jlewis200** [reply]
I did end up finding a workaround.  It involves creating a custom Docker image based off of the so-zeek Dockerfile and installing any additional parsers during the build stage.  I also had to make similar modifications to the so-filebeat Dockerfile to copy in some custom configs to forward the zeek logs through the pipeline.  Finally, I had to create elastic ingest parsers for each protocol.  These can be passed to the container by placing them in the '/opt/so/saltstack/local/salt/elasticsearch/files/ingest/'.  To get everything running you'll have to push the images to the local SO Docker registry and give them a custom tag like 'so-zeek-...'.  To get SO to use the custom images you'll have to edit `/opt/so/saltstack/local/salt/zeek/init.sls` and `/opt/so/saltstack/local/salt/filebeat/init.sls` to pull the images with the custom tags.

**Reply by archery713** [reply]
Thank you so much for the response. I know what the rest of the week will look like now.

**Reply by xfaith** [reply]
Could you provide some sort of install guide for others t follow?

**Reply by archery713** [reply]
> Could you provide some sort of install guide for others t follow?

I am slowly figuring it out yes. I am documenting it for work to begin with so hopefully I will have results within a few days. Currently I made the new Zeek container, installed necessary packages (wget, Cmake, and git). Since the zkg method doesn't work you might not need Python3. I did still install Python3, Python3-pip, and then zkg if I want to customize this in the future more. 

After ./configure for icsnpp-enip I ran make but it is spitting back errors and I'm slowly figuring out the issue. Cmake is 3.24.0-rc4, GCC is 10.2.1. 

Since the container is Debian based updating and installing is easier but still got issues.

Update: I was running on Zeek 3.0.0, there is a 4.2.0, will update if anything changes.

**Reply by archery713** [reply]
Success Update @xfaith 
Zeek container version 4.2.0 with the following installed
> python3, python3-pip, Cmake, git, wget, zkg, libpcap-dev 

Once installed, you should be able to succesfully manually compile incsnpp-enip, and it should show up in "zeek -N"
I will update as I continue down the road to integrating it with Security Onion.

**Reply by archery713** [reply]
@jlewis200 I believe I have configured both my Zeek and Filebeat containers properly. I just have two questions.

1. To receive proper input from the Zeek container, do I only need to edit the filebeat.yml and include the container ID? Or do I need to do more since it is container to container?
2. You mentioned editing "/opt/so/saltstack/local/salt/zeek/init.sls" and "/opt/so/saltstack/local/salt/filebeat/init.sls" to pull the containers with their custom tags. Those files are empty for me. Do you have an example or a resource on how these files should be written? 

Thank you

**Reply by reyesj2** [reply]
@archery713 
You can get the default files from 

/opt/so/saltstack/default/salt/zeek/init.sls

/opt/so/saltstack/local/salt/filebeat/init.sls

**Reply by jlewis200** [reply]
I apologize, I left out the part about copying the default config to the local folder, it's been a few months since I worked on this.  The default configs will in general be found in `/opt/so/saltstack/default/` and can be overridden by placing a custom config in `/opt/so/saltstack/local/`.  So start by copying `/opt/so/saltstack/default/salt/filebeat/init.sls` to  `/opt/so/saltstack/local/salt/filebeat/init.sls` and go from there.  

For filebeat, I had to make a custom filebeat parser for each new zeek parser.  Each new filebeat parser folder should be copied into the `/usr/share/filebeat/module/zeek/` folder of the custom so-filebeat image.  The structure of the filebeat parsers are easy to figure out going off the defaults in `/usr/share/filebeat/module/zeek/`.

You'll also need to edit the global.sls file and include the new zeek modules under the zeek section.  See https://docs.securityonion.net/en/2.3/zeek.html?highlight=custom%20zeek#custom-scripts for an example.  

Also in the global.sls file, there needs to a config for each new filebeat parser, something like this:
```
filebeat:
  config:
    inputs:
      - type: log
        paths:
          - /nsm/zeek/logs/current/[logname].log
        fields:
            module: zeek
            dataset: [logname] 
            category: network
        processors:
        - drop_fields:
            fields: ["source", "prospector", "input", "offset", "beat"]
        fields_under_root: true
        clean_removed: true
        close_removed: false
```

I'm working with my employer to see if we can opensource the whole project, but it's ultimately their decision.
